conn.go

```diff
@@ -20,34 +23,103 @@ var log = logging.Logger("conn")
 
 // ReleaseBuffer puts the given byte array back into the buffer pool,
 // first verifying that it is the correct size
+// TODO: do we need this? I don't see it's used anywhere
 func ReleaseBuffer(b []byte) {
 	log.Debugf("Releasing buffer! (cap,size = %d, %d)", cap(b), len(b))
 	mpool.ByteSlicePool.Put(uint32(cap(b)), b)
 }
 
-// singleConn represents a single connection to another Peer (IPFS Node).
+// singleConn represents a single stream-multipexed connection to another Peer (IPFS Node).
 type singleConn struct {
-	local  peer.ID
-	remote peer.ID
-	maconn tpt.Conn
-	event  io.Closer
+	streamConn smux.Conn
+	tptConn    tpt.Conn
+
+	secSession secio.Session
+
+	event io.Closer
 }
 
-// newConn constructs a new connection
-func newSingleConn(ctx context.Context, local, remote peer.ID, maconn tpt.Conn) (iconn.Conn, error) {
-	ml := lgbl.Dial("conn", local, remote, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
+var _ iconn.Conn = &singleConn{}
+
+// newSingleConn constructs a new connection
+func newSingleConn(ctx context.Context, local, remote peer.ID, privKey ci.PrivKey, tptConn tpt.Conn, pstpt smux.Transport, isServer bool) (iconn.Conn, error) {
+	ml := lgbl.Dial("conn", local, remote, tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+
+	var streamConn smux.Conn
+	var secSession secio.Session
+	if ssc, ok := tptConn.(tpt.SingleStreamConn); ok {
+		var err error
+		// 1. secure the connection
+		secSession, err = setupSecureSession(ctx, local, privKey, ssc)
+		if err != nil {
+			// TODO: return a more specific error code here
+			return nil, err
+		}
+
+		// 2. start stream multipling
+		streamConn, err = pstpt.NewConn(ssc, isServer)
+		if err != nil {
+			return nil, err
+		}
+	} else if msc, ok := tptConn.(tpt.MultiStreamConn); ok {
+		// 1. secure the connection on the first best stream
+		var stream smux.Stream
+		var err error
+		if isServer {
+			stream, err = msc.AcceptStream()
+		} else {
+			stream, err = msc.OpenStream()
+		}
+		if err != nil {
+			return nil, err
+		}
+		defer stream.Close()
+
+		secSession, err = setupSecureSession(ctx, local, privKey, stream)
```

Comment:
Hrm... i'm not sure this is the right way to do this. This is encrypting on top of of the stream muxer, instead of underneath it.

I think that instead of doing it this way, we can just not support private networks and QUIC at the same time. Just error out instead i'm thinking.

cc @Kubuxu @lgierth